### ISSUE TITLE
Fastwiki conflicts with dropfiles

### ISSUE THREAD

**Issue by mr-floppy** [zioth/dokuwiki_fastwiki#36]
Hi,

the plugin dropfiles (https://www.dokuwiki.org/plugin:dropfiles) doesn't let you drag&drop files into the edit box, when fastwiki is activated.

I see you have added a workaround for the plugin imgpaste (https://www.dokuwiki.org/plugin:imgpaste) in the past: https://github.com/zioth/dokuwiki_fastwiki/issues/22

I have tried to copy that exact same fix to dropfiles, but it did not change **anything**.
For what it's worth, I will paste the code here to show how it is **not** done ;)

```js
// file: src/upload.js
jQuery(function () {
    
    var _didInit = false;
    function init() {

		'use strict';

		var $lastKnownCaretPosition = 0; // IE 11 fix
		var $editarea = jQuery('#wiki__text');
		var $filelisting = jQuery('.plugin__filelisting');
		if ((!$editarea.length && !$filelisting.length) || _didInit) {
			return;
		}
		_didInit = true;

		/** UNCHANGED DROPFILES CODE GOES HERE... **/

    }

    init();

    // fastwiki plugin support
    jQuery(window).on('fastwiki:afterSwitch', function(evt, viewMode, isSectionEdit, prevViewMode) {
        if (viewMode == 'edit' || isSectionEdit) {
            init();
        }
    });
});
```

Do you think it is possible to fix this issue in a similar way?

Best regards!

**Comment by zioth** [comment]
The fix might be something like this, in `src/upload.js`. Replace all code starting with `function bootstrapFunctionality` with this:
```
    var _didInit = false;
    function bootstrapFunctionality() {
        enableDragAndDrop($editarea);
        enableDragAndDrop($filelisting);

        if (!_didInit) {
            var widgetTitle = window.LANG.plugins.dropfiles['title:fileUpload'];
            var $widget = jQuery('<div title="' + widgetTitle + '" id="' + UPLOAD_PROGRESS_WIDGET_ID + '"></div>').hide();
            jQuery('body').append($widget);
        }
        _didInit = true;
    }

    function init() {
        $editarea = jQuery('#wiki__text');
        $filelisting = jQuery('.plugin__filelisting');
        if (!$editarea.length && !$filelisting.length) {
            return;
        }

        bootstrapFuntionality();

        $editarea.blur(function () {
            // IE 11 fix
            $lastKnownCaretPosition = $editarea[0].selectionStart;
        });
    }

    init();

    // fastwiki plugin support
    jQuery(window).on('fastwiki:afterSwitch', function(evt, viewMode, isSectionEdit, prevViewMode) {
        if (viewMode == 'edit' || isSectionEdit) {
            init();
        }
    });
```

I'm not certain about this. It might need some fixes for when sectionedit is shown and hidden multiple times.

**Comment by mr-floppy** [comment]
Whow you are fast!
Plus, this seems to work exactly as expected!

Thank you very much!! :)

I will test it over the next couple of days and see, if I run into some problems, but for now I have not noticed anything out of the ordinary.

**Comment by zioth** [comment]
I've fixed a few potential issues and submitted a PR. You can track this ticket to see when it's merged, or encourage the author to merge it :)

https://github.com/cosmocode/dokuwiki-plugin-dropfiles/issues/7

**Comment by mr-floppy** [comment]
Thanks again, you are awesome!!

I have installed the PR manually, but I see they are already working on merging it :)

**Comment by mr-floppy** [comment]
So, I guess the merge failed, because of PHP 7?
I have had a sneak peak just out of curiosity, but did not understand it.

Travis CI says:
```
PHP Fatal error: 
Declaration of DokuWikiTest::setUpBeforeClass() must be compatible with PHPUnit\Framework\TestCase::setUpBeforeClass():
void in /home/travis/build/cosmocode/dokuwiki-plugin-dropfiles/_test/core/DokuWikiTest.php
on line 27
```

I am running `PHP 7.0.33-0ubuntu0.16.04.5` and everything seems fine ;)

**Comment by zioth** [comment]
I saw that. All of their builds are failing, not just this one, and a JS-only change shouldn't have affected a PHP build. If they tell me that they can't merge because of the failed build, I can take a look. Until then, I'll leave it to them.

**Comment by zioth** [comment]
This has been merged into the `dropfiles` development branch. It will be merged into their master branch whenever they feel like it. :)

Closing.
